### DISCUSSION TITLE
Call the onError function or rollback from within onSuccess

### DISCUSSION THREAD

**Discussion by jrcardwell** [TanStack/query#1339] [Q&A] [answered]
Our API posts return a 'message' column that may be blank or contain a warning or error message.  I want to rollback an optimistic mutation if in the OnSuccess function the message is not blank. Is there a way to call OnError or access the (now mutated) original values with which to rollback from within Mutate's OnSuccess ?

**Comment by TkDodo** [comment, marked as answer]
I think the easiest way would be to handle it on the mutation function itself. You can either:
- handle this on the server - send a failure status code with a warning / error instead of a successful response.
- map to an erroneous response on the client in the mutate function: await the actual result from the backend and then transform / throw if it wasn't really successful.

**Reply by jrcardwell** [reply]
Thank you TkDodo - the second option works like a charm.
